### PR TITLE
Fix `stripTrailingHardline` function

### LINKED ISSUE #9734: Extra newline at the end of JavaScript fenced code block with string literal
**Prettier 2.2.0**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEBZAhgJwNYBMIDuUAOiVAAaUBWAzqcSAO0AEDpl5IANCBAA4wAltBrJQWTIQAKWBKJToANgXQBPUTwBGmdGGxwYAZT67BUAObIYmAK5wecALaa4uXK4Ay6CzfTm4AGIQmI7oMEIWyCDoNjAQ3CAAFjCOigDqiYLwNCZgcIZyWYIAblmqUWA0GiBmNHCYMFI65qHIAGZKdTy0AB4AQjp6BobojnAeZnDtnfYgvYZm5opwAIo2EPDTil0gJph1mFGhOPhECXyYZjBpgrgwicgAHAAMPBcQdWk6fFEXcAfFKY8ACO63gTX48miNAAtFA4K5XAlMHBQYIUU0-K0kB1trM6o5BFZbPjFss1hspjiZjwYOhNDc7g8kAAmWk6QSKRYAYQgjmxIH+AFYEjY6gAVenyXE7Yp2ACSUHcsEMYEuAgAgkrDDBVMstnUAL6GoA)
```sh
--parser markdown
```

**Input:**

````markdown
Markdown

```js
"· "
```
````

**Output:**

````markdown
Markdown

```js
"· ";

```

````

**Expected behavior:**

I would expect not to have an extra newline added at the end of the fenced code block. This started happening in prettier v2.1.0, I imagine because of `--embedded-language-formatting`. I couldn’t find any other issue opened about this, at least when searching for issues relating to code blocks.

Trying this with more [JavaScript literals](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEBZAhgJwNYBMIDuUAOiVAAaUBWAzqcSPDTA6ZefRdXVAORMxebSp3a02AjlynDy4qADN0AGxpxZo7qRiYAruuma5PKLuXKNZMTwDaAXUuz5wAL6Pp8gKzvrpAJKKAJZQgTAAnj5aUABy6NGRxqS6ULhwCsFwuLIgADQgEAAOMIHQNMigWJiEAApYCGUoKgToYWV5AEaY6GDYcDAAygXdwQDmyDr6eXAAtu2ZqbgAMuhQI7roI3AAYhCY0+gwxavIIOi6MBC5IAAWMNPKAOrXoXA0Q2Bw-fWhgQBuoWETmAaG0QME1JgYNUuiN9sglKo4HlaAAPABCXR6fX66GmcEWGXhKjUyJoKP6o2UcAAiroIPAiYi8kNMBCTvscPgiFcCphgjAHoFcDBrsgABwABmZVTUDy6BROvNecEwvzgVwAjnT4NDCg1TjQALRQOCZTJXTBwLWBS3QjZwpAIkkgNTTQKM500Sk07Xqx3EpGMdDtQXC0VIABMeR06ECylGAGEINMHSBXp4rro1AAVYMNJ2B376AKpWD9MB8ooAQRS-XCVI9cBcLiAA), this seems to happen only when the code block ends with a string literal with single or double quotes, even if there is a comment after the value.

### PR DESCRIPTION
**Pull request by fisker**
Fixes #9734

- [x] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory)
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/pr-XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

File: src/document/doc-utils.js [hunk 1]
```diff
@@ -193,8 +193,26 @@ function getInnerParts(doc) {
   return parts;
 }
 
-function stripTrailingHardline(doc, withInnerParts = false) {
-  // HACK remove ending hardline, original PR: #1984
+// Remove empty docs
+function cleanDoc(doc) {
+  return mapDoc(doc, (currentDoc) => {
+    const { parts } = currentDoc;
+    if (Array.isArray(parts)) {
```
Comment (thorn0) [inline src/document/doc-utils.js:200]: `fill` has `parts` too and [you can't](https://github.com/prettier/prettier/blob/099c5fb7302021c894e38f53bb09909d45870502/src/document/doc-printer.js#L366-L367) remove individual parts from it that easy.

**Review by thorn0 — CHANGES_REQUESTED** [review]

File: src/document/doc-utils.js (hunk 1 above)
Comment (fisker) [reply src/document/doc-utils.js:200]: Sorry, I didn't think though, empty `parts`  and empty strings in `concat` are safe to remove, right?

Comment (thorn0) [reply src/document/doc-utils.js:200]: Right. Turns out the existing function `normalizeDoc` has this bug too. Why not use this function in this PR, BTW?

Comment (thorn0) [reply src/document/doc-utils.js:200]: #9503 may be caused by this bug in `normalizeDoc` (or may be not).

Comment (fisker) [reply src/document/doc-utils.js:200]: > Right. Turns out the existing function normalizeDoc has this bug too. Why not use this function in this PR, BTW?

I tried it first, not working(forgot how)

**Comment by fisker** [comment]
@thorn0 

I rewrote `stripTrailingHardline`, not it look for `hardlines` anywhere, even in `if-break`.

Add a new `cleanDoc`, the `cleanDoc` 
a) concat `concat` and `string`
b) flat `concat`
c) remove empty string and `concat`


Because side effect of `cleanDoc` , I have to rewrite the embed css part , and it's much simpler now.

I tried to replace `normalizeDoc` with the new one, but seems some tests failed, I guess some `normalizeParts` mean to normalize `parts` of `fill`, I guess we can keep both for now.

**Comment by thorn0** [comment]
A while ago we refactored these `normalize*` functions and unified them without taking into account this trait of `fill`, so we might have broken something. I'll look into this later too.

**Comment by fisker** [comment]
Traverse docs seem not a good solution for this, do you think we should add a new command to do this?

File: src/document/doc-utils.js [hunk 2]
```diff
@@ -180,42 +180,99 @@ function removeLines(doc) {
...
-    parts = lastPart.contents.parts;
+
+  switch (doc.type) {
+    case "concat":
+    case "fill": {
+      const { parts } = doc;
+
+      while (parts.length > 1 && isHardline(...parts.slice(-2))) {
+        parts.pop();
+        parts.pop();
```
Comment (thorn0) [inline src/document/doc-utils.js:201]:
```suggestion
        parts.length -= 2;
```

File: src/document/doc-utils.js [hunk 3]
```diff
@@ -180,42 +180,99 @@ function removeLines(doc) {
...
+      parts[index] = currentDoc;
     }
   }
 
+  if (parts.length === 0) {
+    return "";
+  }
+
+  if (parts.length === 1) {
+    return cleanDocFn(parts[0]);
```
Comment (thorn0) [inline src/document/doc-utils.js:268]: Why call `cleanDocFn` here again? Aren't all the parts already processed at this point?

Comment (fisker) [reply src/document/doc-utils.js:268]: I guess I because I didn't turn `{type: "concat", parts: []}` into `""` at first, I'll check.

Comment (thorn0) [reply src/document/doc-utils.js:267]: Overall, it's probably just me, but it's hard to follow what's going on in this function as compared with `normalizeDoc`, which looks much simpler. Can we probably just use `normalizeDoc` with an extra parameter to protect `fill`?

Comment (fisker) [reply src/document/doc-utils.js:278]: `normalizeDoc` didn't remove empty `concat`, so we can't locate hardLine.

**Comment by fisker** [comment]
Funny thing, I found there is another function `flattenDoc` in `doc-debug.js`.

Here is the result, running on '/index.js'

<details>
<summary>Original</summary>

```json
{
  "type": "concat",
  "parts": [
    {
      "type": "concat",
      "parts": [
        {
          "type": "concat",
          "parts": [
            {
              "type": "concat",
              "parts": [
                {
                  "type": "concat",
                  "parts": [
                    {
                      "type": "concat",
                      "parts": [
                        "\"use strict\""
                      ]
                    }
                  ]
                },
                ";",
                {
                  "type": "concat",
                  "parts": [
                    {
                      "type": "line",
                      "hard": true
                    },
                    {
                      "type": "break-parent"
                    }
                  ]
                },
                {
                  "type": "concat",
                  "parts": [
                    {
                      "type": "line",
                      "hard": true
                    },
                    {
                      "type": "break-parent"
                    }
                  ]
                },
                {
                  "type": "concat",
                  "parts": [
                    {
                      "type": "concat",
                      "parts": [
                        {
                          "type": "concat",
                          "parts": [
                            {
                              "type": "concat",
                              "parts": [
                                {
                                  "type": "concat",
                                  "parts": [
                                    {
                                      "type": "group",
                                      "contents": {
                                        "type": "concat",
                                        "parts": [
                                          {
                                            "type": "concat",
                                            "parts": [
                                              {
                                                "type": "concat",
                                                "parts": [
                                                  {
                                                    "type": "concat",
                                                    "parts": [
                                                      {
                                                        "type": "concat",
                                                        "parts": [
                                                          "module",
                                                          "",
                                                          ""
                                                        ]
                                                      }
                                                    ]
                                                  },
                                                  {
                                                    "type": "concat",
                                                    "parts": [
                                                      "",
                                                      ".",
                                                      {
                                                        "type": "concat",
                                                        "parts": [
                                                          {
                                                            "type": "concat",
                                                            "parts": [
                                                              "exports",
                                                              "",
                                                              ""
                                                            ]
                                                          }
                                                        ]
                                                      }
                                                    ]
                                                  }
                                                ]
                                              }
                                            ]
                                          },
                                          {
                                            "type": "concat",
                                            "parts": [
                                              " ",
                                              "="
                                            ]
                                          },
                                          {
                                            "type": "concat",
                                            "parts": [
                                              " ",
                                              {
                                                "type": "concat",
                                                "parts": [
                                                  {
                                                    "type": "concat",
                                                    "parts": [
                                                      "",
                                                      {
                                                        "type": "concat",
                                                        "parts": [
                                                          {
                                                            "type": "concat",
                                                            "parts": [
                                                              "require",
                                                              "",
                                                              ""
                                                            ]
                                                          }
                                                        ]
                                                      },
                                                      "",
                                                      "",
                                                      {
                                                        "type": "concat",
                                                        "parts": [
                                                          "(",
                                                          {
                                                            "type": "concat",
                                                            "parts": [
                                                              {
                                                                "type": "concat",
                                                                "parts": [
                                                                  "\"./src/index\""
                                                                ]
                                                              }
                                                            ]
                                                          },
                                                          ")"
                                                        ]
                                                      }
                                                    ]
                                                  }
                                                ]
                                              }
                                            ]
                                          }
                                        ]
                                      },
                                      "break": false
                                    }
                                  ]
                                },
                                ";"
                              ]
                            }
                          ]
                        }
                      ]
                    }
                  ]
                },
                "",
                {
                  "type": "concat",
                  "parts": [
                    {
                      "type": "line",
                      "hard": true
                    },
                    {
                      "type": "break-parent"
                    }
                  ]
                }
              ]
            }
          ]
        }
      ]
    }
  ]
}
```

</details>

<details>
<summary>CleanDoc</summary>

```json
{
  "type": "concat",
  "parts": [
    "\"use strict\";",
    {
      "type": "line",
      "hard": true
    },
    {
      "type": "break-parent"
    },
    {
      "type": "line",
      "hard": true
    },
    {
      "type": "break-parent"
    },
    {
      "type": "group",
      "contents": "module.exports = require(\"./src/index\")",
      "break": false
    },
    ";",
    {
      "type": "line",
      "hard": true
    },
    {
      "type": "break-parent"
    }
  ]
}
```

</details>


<details>
<summary>flattenDoc</summary>

```json
{
  "type": "concat",
  "parts": [
    "\"use strict\"",
    ";",
    {
      "type": "line",
      "hard": true
    },
    {
      "type": "break-parent"
    },
    {
      "type": "line",
      "hard": true
    },
    {
      "type": "break-parent"
    },
    {
      "type": "group",
      "contents": {
        "type": "concat",
        "parts": [
          "module",
          ".",
          "exports",
          " ",
          "=",
          " ",
          "require",
          "(",
          "\"./src/index\"",
          ")"
        ]
      },
      "break": false
    },
    ";",
    {
      "type": "line",
      "hard": true
    },
    {
      "type": "break-parent"
    }
  ]
}
```

</details>

File: src/document/doc-utils.js (hunk 3 above)
Comment (thorn0) [reply src/document/doc-utils.js:278]: If I correctly understand what you mean, originally it did that:

https://github.com/prettier/prettier/pull/8496/files#diff-751b7c98ecb8b8f58f17f22b778c467270f4d0a4fd6b72a262c71a8a243888b6L860-L862

But that was somehow lost in refactoring. Try adding it back?

Comment (fisker) [reply src/document/doc-utils.js:278]: Example

```json
{
	type:"concat",
	parts: [
		{
			type: 'concat'
			[
				{type: "concat", parts: [...HARD_LINE]}
			],
		}
		// A empty string
		{
			type: 'concat',
			parts: ['']
		}
	]
}
```

Comment (thorn0) [reply src/document/doc-utils.js:278]: Looks like exactly what those removed lines I linked to were for.

**Comment by thorn0** [comment]
As for `flattenDoc` and `--debug-print-doc`, I thinks for debug purposes, probably it's better not to normalize too much. Also it traverses `expandedStates`.

File: src/document/doc-utils.js (hunk 3 above)
Comment (fisker) [reply src/document/doc-utils.js]: > But that was somehow lost in refactoring. Try adding it back?

I  prefer use the new one. It do much more normalize than the old one, we can extend it to let it work for `fill` if we want replace old one later.

More clean, more easy for finding these new lines.

**Comment by fisker** [comment]
@sosukesuzuki  I remember you have a playground for docs? What do you think the new cleaned doc https://github.com/prettier/prettier/pull/9736#issuecomment-733760548

File: src/document/doc-utils.js [hunk 4]
```diff
@@ -180,42 +180,127 @@ function removeLines(doc) {
...
+      currentDoc += nextDoc;
+    }
+    // Concat continuous `concat`
+    else if (isConcat(currentDoc) && isConcat(nextDoc)) {
+      parts.splice(index + 1, 1);
+      currentDoc.parts.push(...nextDoc.parts);
+      currentDoc = cleanDocFn(currentDoc);
+    }
+    // Concat `concat` and string
+    else if (isConcat(currentDoc) && typeof nextDoc === "string") {
```
Comment (thorn0) [inline src/document/doc-utils.js:271]: Can you please add a test for what is achieved with these manipulations? As this function is passed to `mapDoc`, which by itself is recursive and depth-first, I'm not sure about the neccessity of these recursive calls here.

File: src/document/doc-utils.js [hunk 5]
```diff
@@ -180,42 +180,127 @@ function removeLines(doc) {
...
-        type: "concat",
-        parts: doc.parts.slice(0, -1).concat(stripTrailingHardline(lastPart)),
-      };
+  if (!isConcat(doc)) {
+    return doc;
+  }
+
+  const { parts } = doc;
+  for (let index = parts.length - 1; index >= 0; index--) {
+    let currentDoc = parts[index];
```
Comment (thorn0) [inline src/document/doc-utils.js:255]: `currentPart` and `nextPart` could be better names IMHO.

**Comment by thorn0** [comment]
As a completely different and possibly less hacky solution for the problem at hand, I wonder if it's possiblet to add a new command for the printer that would "unprint" the last printed line break it it's the last thing that was printed. (Just a thought.)

**Comment by fisker** [comment]
I did try it, when checking the new command in print, it's too late, the docs before already decided "break" or "flat".

File: src/document/doc-utils.js (hunk 4 above)
Comment (fisker) [reply src/document/doc-utils.js:271]: Yes, there are recursive calls, but the all needed. Because the inner "concat" already "cleaned", but it's updated, have to clean it again , I can't figure a better way.

Comment (thorn0) [reply src/document/doc-utils.js:271]: Let's find a better way to do this. These nested recursions look suspiciously like they're O(n**2).

Comment (fisker) [reply src/document/doc-utils.js]: I may found something, I forgot to take "concat.parts" out, will try.

File: src/document/doc-utils.js [hunk 6]
```diff
@@ -180,42 +180,117 @@ function removeLines(doc) {
...
+    }
+    case "align":
+    case "indent":
+    case "group":
+    case "line-suffix":
+      doc.contents = stripDocTrailingHardlineFromDoc(doc.contents);
+      break;
+    case "if-break":
+      doc.breakContents = stripDocTrailingHardlineFromDoc(doc.breakContents);
+      doc.flatContents = stripDocTrailingHardlineFromDoc(doc.flatContents);
```
Comment (thorn0) [inline src/document/doc-utils.js:219]: Why do you not use `mapDoc` here?

File: src/document/doc-utils.js [hunk 7]
```diff
@@ -180,42 +180,117 @@ function removeLines(doc) {
...
+
+const isConcat = (doc) => doc && doc.type === "concat";
+function cleanDocFn(doc) {
+  switch (doc.type) {
+    case "align":
+    case "indent":
+    case "group":
+    case "line-suffix":
+      if (!doc.contents) {
+        return "";
```
Comment (thorn0) [inline src/document/doc-utils.js:239]: This can break conditional groups.

Comment (fisker) [reply src/document/doc-utils.js:239]: Yes, found it in #9784, I'm testing

File: src/document/doc-utils.js (hunk 6 above)
Comment (fisker) [reply src/document/doc-utils.js:219]: It only strip from last item in "parts". `mapDoc` can't know a element is last.

Correct me if I'm wrong

File: src/document/doc-utils.js [hunk 8]
```diff
@@ -180,42 +180,135 @@ function removeLines(doc) {
...
       }
+      break;
+    case "if-break":
+      if (!doc.flatContents && !doc.breakContents) {
+        return "";
+      }
+      break;
+  }
```
Comment (fisker) [inline src/document/doc-utils.js:265]: @thorn0 I pushed some changes, I think there is no problem in `concat` part, please review these lines above, are they safe?

File: src/document/doc-utils.js (hunk 6 above)
Comment (thorn0) [reply src/document/doc-utils.js:219]: Right, you're correct. I think it should also traverse `extendedStates`.

File: src/document/doc-utils.js (hunk 8 above)
Comment (thorn0) [reply src/document/doc-utils.js:265]: Please make this function pure. Let's avoid mutating objects when it's not necessary.

File: src/document/doc-utils.js (hunk 6 above)
Comment (thorn0) [reply src/document/doc-utils.js:219]: Re `extendedStates`. In theory, it should be done, but in practice the trailing hard line we want to remove is probably never rendered by a `conditionalGroup`, so let's skip this for now unless you feel like researching this.

File: src/document/doc-utils.js [hunk 9]
```diff
@@ -180,42 +180,132 @@ function removeLines(doc) {
...
+    case "group":
+      // Remove nested only group
       if (
-        lastPart.parts.length === 2 &&
-        lastPart.parts[0].hard &&
-        lastPart.parts[1].type === "break-parent"
+        doc.contents.type === "group" &&
+        doc.contents.id === doc.id &&
+        doc.contents.break === doc.break &&
+        doc.contents.expandedStates === doc.expandedStates
```
Comment (thorn0) [inline src/document/doc-utils.js:245]: I don't understand this part. I thought you'd just need to replace `if (!doc.contents)` with `if (!doc.contents && !doc.expandedStates)`.

Comment (fisker) [reply src/document/doc-utils.js:245]: For hardline searching it's not needed, but I try to make it safe  to reduce nesting groups, for later use.

You can format `.codecov.yml`, I'm shocked, JSON.stringify the doc  with 2 spaces, it's 20mb...

File: src/document/doc-utils.js (hunk 8 above)
Comment (fisker) [reply src/document/doc-utils.js:262]: I read the code again, I think it's pure, I only mutate parts of "concat", I thought it will be faster, and parts already a copy in `mapDoc`.

File: src/document/doc-utils.js (hunk 9 above)
Comment (fisker) [reply src/document/doc-utils.js:245]: Format this 128b `.codecov.yml` file,

Stringify the doc with

```js
	JSON.stringify(
      doc,
      (key, value) => (key === "expandedStates" ? undefined : value),
      2
    )
```

Original: 1803KB, with this code 266KB, without this 392KB

File: src/document/doc-utils.js (hunk 8 above)
Comment (fisker) [reply src/document/doc-utils.js:262]: `stripTrailingHardline` was not pure, https://github.com/prettier/prettier/pull/9736/commits/fdbb34cfbdbe80dfc826b19f75d1092a3b16273b

Comment (thorn0) [reply src/document/doc-utils.js:264]: I mean `cleanDoc`. But it's better to make `stripTrailingHardline` pure too. Pure functions are cleaner code.

Comment (fisker) [reply src/document/doc-utils.js:264]: Do you mean I should copy `parts` again in `cleanDoc`? Is that really necessary.

File: src/document/doc-utils.js (hunk 9 above)
Comment (thorn0) [reply src/document/doc-utils.js:247]: It's better to measure and compare JSON without spaces.

But back to my question. Yes, let's keep hardline searching out of the equation and think of this as of a generic normalizing function. Is this condition ever true: `doc.contents.id === doc.id`? Aren't ids supposed to be unique? What's wrong with the replacement I mentioned in https://github.com/prettier/prettier/pull/9736#discussion_r531718688?

File: src/document/doc-utils.js (hunk 8 above)
Comment (thorn0) [reply src/document/doc-utils.js:264]: As for what is faster, I doubt `splice` is faster. Because it needs to shift all the subsequent elements. I wouldn't use `splice` like that without having significant benchmark numbers proving that this really makes sense.

File: src/document/doc-utils.js (hunk 9 above)
Comment (fisker) [reply src/document/doc-utils.js:247]: group.id most time just "undefined", if there are not equal, it can't be removed, it will effect ifBreak, only if it's "undefined", removing the inner one will not effect the ifBreak, because they both decided by the parent group.

I didn't learn too much about docs, but I think this is how they works?

Comment (fisker) [reply src/document/doc-utils.js:247]: > It's better to measure and compare JSON without spaces.

Original: 60KB, remove group 26KB, without removing group 34KB

Comment (thorn0) [reply src/document/doc-utils.js:247]: Indeed, I didn't think about the `undefined === undefined` case. `!doc.id && !doc.contents.id` looks like a cleaner way to express this logic. Anyway, it's also definitely safe to remove a group if it has empty `contents` (`""` or `concat([])`), no `id`, and no `expandedStates`.

Comment (fisker) [reply src/document/doc-utils.js:247]: > !doc.id && !doc.contents.id

 Won't work if they have same symbol id, they can be removed too. :)

I tried to remove empty `contents` group, but test failed, maybe because I didn't check `id` and `expandedStates`

Comment (thorn0) [reply src/document/doc-utils.js:247]: "they have same symbol id" - Is this really possible? Sounds like a bug in the printer.

Yes, you have to check all the properties like I described above (note that I edited that comment).

Comment (fisker) [reply src/document/doc-utils.js:247]: Don't know why yaml printer use so many groups.

Comment (fisker) [reply src/document/doc-utils.js:247]: > "they have same symbol id" - Is this really possible? Sounds like a bug in the printer.


I think yes, but maybe not nested, in js printer there is getXXXGroupId function, it's cached, I'm in bed now, you can find it.

File: src/document/doc-utils.js [hunk 10]
```diff
@@ -180,41 +180,123 @@ function removeLines(doc) {
...
+    const previousPart = newParts[newParts.length - 1];
+    const isPreviousPartString = typeof previousPart === "string";
+    const [currentPart, ...rest] = isConcat(part) ? part.parts : [part];
+    if (isPreviousPartString && typeof currentPart === "string") {
+      newParts[newParts.length - 1] += currentPart;
+    } else {
+      newParts.push(currentPart);
     }
+    newParts.push(...rest);
   }
```
Comment (thorn0) [inline src/document/doc-utils.js:287]: So, this is the same thing as `normalizeParts` now, I suggest we move your implementation to that function and call it here.

Comment (fisker) [reply src/document/doc-utils.js]: Feels not safe for this part https://github.com/prettier/prettier/blob/e15e6112b617ecadaae76fc624378231146fb0e9/src/language-html/printer-html.js#L387-L395

Comment (thorn0) [reply src/document/doc-utils.js]: You're right, I didn't think about that. Let's don't do this then.

Comment (fisker) [reply src/document/doc-utils.js]: Reverted.

**Review by thorn0 — APPROVED** [review]
I really don't like having 2 very similar functions with similar names `cleanDoc` and `normalizeDoc`, but hopefully it'll be refactored later. Please add a short comment explaining the difference between these functions.

Otherwise, looks good.
